src/ol-ext/format/featurehash.js

```diff
@@ -0,0 +1,940 @@
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writePointGeometry_ = function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.Point);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var end = flatCoordinates.length;
+  return 'p(' + this.encodeCoordinates_(flatCoordinates, stride, 0, end) + ')';
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @return {string} Encoded geometry.
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writeMultiPointGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPoint);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var end = flatCoordinates.length;
+  return 'P(' + this.encodeCoordinates_(flatCoordinates, stride, 0, end) + ')';
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @return {string} Encoded geometry.
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writePolygonGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.Polygon);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var ends = geometry.getEnds();
+  var linearRingCount = ends.length;
+  var offset = 0;
+  var textArray = ['a('];
+  for (var i = 0; i < linearRingCount; ++i) {
+    // skip the "closing" point
+    var end = ends[i] - stride;
+    var text = this.encodeCoordinates_(flatCoordinates, stride, offset, end);
+    if (i !== 0) {
+      textArray.push('\'');
+    }
+    textArray.push(text);
+    offset = ends[i];
+  }
+  textArray.push(')');
+  return textArray.join('');
+};
+
+
+/**
+ * @param {ol.geom.Geometry} geometry Geometry.
+ * @param {olx.format.WriteOptions=} opt_options Write options.
+ * @return {string} Encoded geometry.
+ * @this {ngeo.format.FeatureHash}
+ * @private
+ */
+ngeo.format.FeatureHash.writeMultiPolygonGeometry_ =
+    function(geometry, opt_options) {
+  goog.asserts.assertInstanceof(geometry, ol.geom.MultiPolygon);
+  var flatCoordinates = geometry.getFlatCoordinates();
+  var stride = geometry.getStride();
+  var endss = geometry.getEndss();
+  var polygonCount = endss.length;
+  var offset = 0;
+  var textArray = ['A'];
+  for (var i = 0; i < polygonCount; ++i) {
+    textArray.push('(');
+    var ends = endss[i];
+    var linearRingCount = ends.length;
+    for (var j = 0; j < linearRingCount; ++j) {
+      // skip the "closing" point
+      var end = ends[j] - stride;
+      var text = this.encodeCoordinates_(flatCoordinates, stride, offset, end);
+      if (j !== 0) {
+        textArray.push('\'');
+      }
+      textArray.push(text);
+      offset = ends[j];
+    }
```

Comment:
Cannot the linearRing loop be factorized? It is used similarly in `writePolygonGeometry_`.